### PR TITLE
Uploaded sample project for parallel with testNG

### PR DESCRIPTION
**Pull request by thaitrongnghi**
@khanhdodang , please help to review

### CODE REVIEW

File: ParallelWithTestNG/testng.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+<suite thread-count="2" name="Kobiton Parallel Suite" parallel="tests" >
+    <parameter name="kobitonURL"
+               value="https://YOUR_USERNAME:YOUR_API_KEY@api.kobiton.com/wd/hub"/> <!-- Ex: username:apiKey@api-test.kobiton.com/wd/hub -->
```
Comment (khanhdodang) [inline ParallelWithTestNG/testng.xml]: Change from "api-test" tp api

File: ParallelWithTestNG/src/main/java/WebTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    public static Map<String,String> testngParams;
+    String kobitonURL;
+    RemoteWebDriver driver = null;
+
+    @BeforeTest
+    @Parameters({"platform"})
+    public void Setup(ITestContext testContext, String platform) throws MalformedURLException {
+        testngParams = testContext.getCurrentXmlTest().getAllParameters();
+        kobitonURL = testngParams.get("kobitonURL");
+        driver = new RemoteWebDriver(new URL(kobitonURL), this.generateDesiredCaps(platform));
```
Comment (khanhdodang) [inline ParallelWithTestNG/src/main/java/WebTest.java]: Handle ` URL(kobitonURL)` as String

File: ParallelWithTestNG/src/main/java/WebTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+
+    @Test
+    public void testKobitonWebsite() throws MalformedURLException, InterruptedException {
+        driver.get("http://google.com");
+        WebElement searchFieldE = driver.findElementByXPath("//input[@name='q']");
+        searchFieldE.sendKeys("Kobiton");
+        searchFieldE.sendKeys(Keys.ENTER);
+        if (driver.findElementsByXPath("//div/span[text()='Looking for results in English?']").size() != 0) {
```
Comment (khanhdodang) [inline ParallelWithTestNG/src/main/java/WebTest.java]: Change to check the title

File: ParallelWithTestNG/src/main/java/WebTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.testng.Assert;
+import org.testng.annotations.*;
+import org.testng.ITestContext;
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import java.util.Map;
+
+public class WebTest {
+    public static Map<String,String> testngParams;
```
Comment (khanhdodang) [inline ParallelWithTestNG/src/main/java/WebTest.java]: Have a space for `Map<String,String>`

File: ParallelWithTestNG/src/main/java/WebTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        searchFieldE.sendKeys("Kobiton");
+        searchFieldE.sendKeys(Keys.ENTER);
+        if (driver.findElementsByXPath("//div/span[text()='Looking for results in English?']").size() != 0) {
+            driver.findElementByXPath("//div/a[text()='Change to English']").click();
+        }
+        String message = driver.findElementByXPath("//a[@href='https://kobiton.com/']").getText();
+        Assert.assertTrue(message.contains("Kobiton"));
+    }
+
+    public static final DesiredCapabilities generateDesiredCaps(String platform) {
```
Comment (khanhdodang) [inline ParallelWithTestNG/src/main/java/WebTest.java]: Remove final

**Comment by khanhdodang** [comment]
Hi a @duylam Am fine with this PR. Please help me review it.

File: ParallelWithTestNG/pom.xml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (duylam) [inline ParallelWithTestNG/pom.xml]: Why not put in `java` folder ?

Comment (duylam) [reply ParallelWithTestNG/pom.xml]: Should update `java/README.md` for how to execute this test: how to build, how to run, how to provide api key, etc.

File: ParallelWithTestNG/src/main/java/WebTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
+/**
+ * Created by Nghi Thai on 12/29/17.
```
Comment (duylam) [inline ParallelWithTestNG/src/main/java/WebTest.java]: Remove the author comment

File: ParallelWithTestNG/src/main/java/WebTest.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        searchFieldE.sendKeys("Kobiton");
+        searchFieldE.sendKeys(Keys.ENTER);
+        String message = driver.getTitle();
+        Assert.assertTrue(message.contains("Kobiton"));
+    }
+
+    public static DesiredCapabilities generateDesiredCaps(String platform) {
+        DesiredCapabilities capabilities = new DesiredCapabilities();
+        String browserName = "chrome";
+        String deviceName = "Galaxy";
```
Comment (duylam) [inline ParallelWithTestNG/src/main/java/WebTest.java]: Put the deviceName as param in `testng.xml`

File: ParallelWithTestNG/src/main/java/WebTest.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import org.testng.annotations.*;
+import org.testng.ITestContext;
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import java.util.Map;
+
+public class WebTest {
+    public static Map<String, String> testngParams;
+    String kobitonURL;
+    RemoteWebDriver driver = null;
```
Comment (ledinhphuong) [inline ParallelWithTestNG/src/main/java/WebTest.java]: I saw in our sample using `RemoteWebDriver` (It's too generic). 
* Refer to use as Appium sample code: `AppiumDriver driver = null;`
* Refactor other classes too.

File: ParallelWithTestNG/pom.xml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.testng</groupId>
+            <artifactId>testng</artifactId>
+            <version>6.10</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-java</artifactId>
```
Comment (ledinhphuong) [inline ParallelWithTestNG/pom.xml]: * Refer to use `java-client` of Appium for consistency with other classes.
* Do we really need to add the new folder structure (new `ParallelWithTestNG` folder) and switch to use maven while we are supporting Gradle? And it's nice to support maven for this project later.

File: ParallelWithTestNG/src/main/java/WebTest.java (hunk 8 above)
Comment (ledinhphuong) [reply ParallelWithTestNG/src/main/java/WebTest.java]: Put more parameters: `deviceName`, `browserName` in testng.xml file

**Comment by thaitrongnghi** [comment]
@duylam , @ledinhphuong , @khanhdodang 
Thanks all for the valuable comments! There are a few changes, let's help take a look at this, please.

File: java/ParallelWithTestNG/src/main/java/WebTest.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    public void testKobitonWebsite() {
+        driver.get("http://google.com");
+        WebElement searchFieldE = driver.findElementByXPath("//input[@name='q']");
+        searchFieldE.sendKeys("Kobiton");
+        searchFieldE.sendKeys(Keys.ENTER);
+        String message = driver.getTitle();
+        Assert.assertTrue(message.contains("Kobiton"));
+    }
+
+    public static DesiredCapabilities generateDesiredCaps(String platform, String deviceName, String browserName) {
```
Comment (khanhdodang) [inline java/ParallelWithTestNG/src/main/java/WebTest.java]: `generateDesiredCaps` -> `generatedDesiredCaps`

File: java/ParallelWithTestNG/src/main/java/WebTest.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                driver.quit();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Test
+    public void testKobitonWebsite() {
+        driver.get("http://google.com");
+        WebElement searchFieldE = driver.findElementByXPath("//input[@name='q']");
```
Comment (khanhdodang) [inline java/ParallelWithTestNG/src/main/java/WebTest.java]: `searchFieldE` -> `searchEle`

File: java/ParallelWithTestNG/src/main/java/WebTest.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+import org.testng.ITestContext;
+import org.openqa.selenium.Keys;
+import java.util.Map;
+
+public class WebTest {
+    AppiumDriver driver = null;
+
+    @BeforeTest
+    @Parameters({"platform", "deviceName", "browserName"})
+    public void Setup(ITestContext testContext, String platform, String deviceName, String browserName) throws MalformedURLException {
```
Comment (khanhdodang) [inline java/ParallelWithTestNG/src/main/java/WebTest.java]: `platform` -> `platformName`

Comment (thaitrongnghi) [reply java/ParallelWithTestNG/src/main/java/WebTest.java]: fixed

File: java/ParallelWithTestNG/src/main/java/WebTest.java (hunk 12 above)
Comment (thaitrongnghi) [reply java/ParallelWithTestNG/src/main/java/WebTest.java]: it's fine for me, no need to fix

File: java/ParallelWithTestNG/src/main/java/WebTest.java (hunk 11 above)
Comment (thaitrongnghi) [reply java/ParallelWithTestNG/src/main/java/WebTest.java]: fixed platformName, but generateDesiredCaps is still fine for me

**Comment by duylam** [comment]
Thanks @thaitrongnghi @khanhdodang and @ledinhphuong

**Pull request merged**